### PR TITLE
[feature/constraints] OS constraint

### PR DESCRIPTION
**Pull request by vlada-shubina**
### Problem
https://github.com/dotnet/templating/issues/3107 - OS constraint

TODO: 
- more tests
- localization

### Checks:
- [ ] Added unit tests
- [ ] Added `#nullable enable` to all the modified files [?](https://github.com/dotnet/templating/wiki/Contributing#coding-style)

### CODE REVIEW

**Comment by vlada-shubina** [comment]
@baronfel @JanKrivanek looking for the feedback here: do we need to have more features here? 
- support other OS
- support exact names (Ubuntu, Alpine ...)
- support OS versions/version range?

As of now only `Windows`, `Linux`, `OSX` are supported (mapped to corresponding `OSPlatform`).

File: src/Microsoft.TemplateEngine.Edge/Constraints/OSConstraint.cs [hunk 1]
```diff
@@ -43,22 +51,82 @@ internal OSConstraint(IEngineEnvironmentSettings environmentSettings, ITemplateC
...
+
+                if (token!.Type == JTokenType.String)
+                {
+                    return new[] { Parse(token.Value<string>()) };
+                }
+                else if (token is JArray jArray)
+                {
+                    IEnumerable<string?> values = jArray.Values<string>();
+                    List<OSPlatform> readValues = new List<OSPlatform>();
+                    foreach (string? value in values)
```
Comment (JanKrivanek) [inline src/Microsoft.TemplateEngine.Edge/Constraints/OSConstraint.cs:103]: Does the ConstraintManager validate the constraint arg schema validity? I'm just wondering what happens if array of non-strings would lead to meaningful error message.

In any case - it is probably worth to error out on empty array

**Review by JanKrivanek — APPROVED** [review]

File: src/Microsoft.TemplateEngine.Edge/Constraints/OSConstraint.cs [hunk 2]
```diff
@@ -16,6 +17,13 @@ namespace Microsoft.TemplateEngine.Edge.Constraints
 {
     internal class OSConstraintFactory : ITemplateConstraintFactory
     {
+        private static readonly Dictionary<string, OSPlatform> _platformMap = new Dictionary<string, OSPlatform>(StringComparer.OrdinalIgnoreCase)
+        {
+            { "Windows",  OSPlatform.Windows },
+            { "Linux",  OSPlatform.Linux },
+            { "OSX",  OSPlatform.OSX }
+        };
```
Comment (JanKrivanek) [inline src/Microsoft.TemplateEngine.Edge/Constraints/OSConstraint.cs:25]: I'm thinking whether `OSPlatform` is good enough for target use cases.
Looking into maui codebase, they do not use it at all (outside of some tests). In their codnitions (e.g. [`OnPlatfrom`](https://github.com/dotnet/maui/blob/f2a42f6aadd018144f191ccb6107734f0904513a/src/Controls/src/Core/OnPlatform.cs#L8)), they rely on [`Device`](https://github.com/dotnet/maui/blob/f2a42f6aadd018144f191ccb6107734f0904513a/src/Controls/src/Core/Device.cs) or it's replacement [`DevicePlatform`](https://github.com/dotnet/maui/blob/4b7dd8acaf2b0040a6a07e71fc140505ee68b7ac/src/Essentials/src/Types/DevicePlatform.shared.cs) (which misses GTK and WPF, but there is probably good reason for that).

**tl;dr;:** We should probably use same level of categorization as Maui codebase or if we use lower level of detail then we should check with them if it's good enough for their use cases.

**Review by JanKrivanek — CHANGES_REQUESTED** [review]
I admit I have no insight and understanding of Maui use cases and requirements for our constraints feature. But I'm worried whether this implementation can fullfill them

File: src/Microsoft.TemplateEngine.Edge/Constraints/OSConstraint.cs (hunk 2 above)
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Edge/Constraints/OSConstraint.cs:25]: I would avoid building on top of MAUI classes (in general).
`OperatingSystem` seems to be a better alternative (https://docs.microsoft.com/en-us/dotnet/api/system.operatingsystem?f1url=%3FappId%3DDev16IDEF1%26l%3DEN-US%26k%3Dk(System.OperatingSystem)%3Bk(SolutionItemsProject)%3Bk(DevLang-csharp)%26rd%3Dtrue&view=net-6.0)

Comment (JanKrivanek) [reply src/Microsoft.TemplateEngine.Edge/Constraints/OSConstraint.cs:25]: After online talk with @vlada-shubina, it seems that `OSPlatform` should actually suffice here - as we are looking at design time constraint, not a run time constraint.

**Review by JanKrivanek — APPROVED** [review]

**Comment by baronfel** [comment]
Agree with the OperatingSystem-based approach. We don't need to handle OS versions, there's no concrete use case for that that I can see. I'd just make sure that case-insensitive comparisons are handled and then this is good to go IMO!

File: src/Microsoft.TemplateEngine.Edge/Constraints/OSConstraint.cs (hunk 1 above)
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Edge/Constraints/OSConstraint.cs:102]: Added the check for empty array. 
There is no validation against schema at all unfortunately, it's part of https://github.com/dotnet/templating/issues/2594.

**Comment by vlada-shubina** [comment]
Opening for final review, now targeting main.

I decided to stay with `OSPlatform` as it has definition for various OS Platforms, in contrary to `OperatingSystem` class. Both are supported in .NET Framework and .NET (Core), so no preference here.
